Adds repressurizing grenades

lordpidey opened the pull request:
:cl: Lordpidey
add: Centcom will now issue repressurizing grenades to Atmospheric personel.  These grenades when used will create numerous short lived portals that will siphon gases from a backwater orphan filled planet, and re-pressurize the room.
/:cl:

So, these grenades when thrown will spawn a bunch of infinite, but short lived air sources.  They are quite useful at refilling areas once the breach has been sealed.

Commit:
Adds repressurizing grenades

FantasticFwoosh commented:
Urgh, i don't get why we couldn't just make a atmos leaf blower that converts chems into gases to do this instead, ask chemistry for a few beakers of oxygen, load it in and spray it around.

Ratvar commented:
@FantasticFwoosh code it and have it? Go do code thing

CPTANT commented:
Meh, I feel there has been some significant powercreep of atmos equipment lately.

CPTANT commented:
If I read it correctly the portals will last for 60 seconds and output infinite gas at 1 atmosphere? 

What is the point of even trying to keep the air distribution network intact if you can solve all your problems with a single grenade?

pubby commented:
Hm. Why these over canisters?

lordpidey commented:
That is a correct observation, @CPTANT.

However, the station gets TWO of these per shift.  They're designed to fix ONE problem each, regardless of size.

If a traitor blows up medbay, even after it gets patched up, it's still going to be inhospitable.  And like hell anyone is going to repair the distribution network during that, so only vents from one direction will be restoring air.

This is meant to be a solution for situations like this, but only once or twice per shift.

Cyberboss reviewed (changes requested):
I really don't like the snowflake canister implementation

Commit:
Changes a list declaration to use a copied define.

lordpidey commented:
Okay, I'll try to think of a better way than a canister.  I figured since it already had all of the functionality as I wanted, it would be best.  Already has good air mix, outputs air, won't output air beyond 1x atm of pressure.

lordpidey commented:
Closing this for now, I'm gearing up for hurricane evacuation, and I might be gone for a week.

Cyberboss commented:
Come back safe!

On Sep 5, 2017 5:38 PM, "Mike Long" <notifications@github.com> wrote:

> Closed #30391 <https://github.com/tgstation/tgstation/pull/30391>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/30391#event-1235751395>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHyweuIzoa4_Qxd7Zy-MIT_UYmrYY6XNks5sfb9qgaJpZM4PLV7E>
> .
>

lordpidey commented:
Okay, I'm safe now, and in a hotel with internet, so I'll resume working on this.  I'll find a way to do this without cannister replacement.

Commit:
Changes portal from being a cannister to being a gas miner.

Commit:
Fixes files I reverted changes to, to ACTUALLY revert all changes

Cyberboss commented on code/modules/atmospherics/machinery/other/miner.dm:
```diff
@@ -13,29 +12,31 @@
 	density = FALSE
 	resistance_flags = INDESTRUCTIBLE|ACID_PROOF|FIRE_PROOF
-	var/spawn_id = null
+	var/list/gas_mixture = list()
```
gas_mixture**s**

lordpidey replied:
It doesn't make sense for mixture to be plural.

The miner has one mixture, which has one or more gasses.

Cyberboss commented on code/modules/atmospherics/machinery/other/miner.dm:
```diff
@@ -122,7 +128,7 @@
 	check_operation()
 	if(active && !broken)
-		if(isnull(spawn_id))
+		if(isnull(gas_mixture) && gas_mixture.len)
```
LAZYLEN(gas_mixture)

Cyberboss commented on code/modules/atmospherics/machinery/other/miner.dm:
```diff
@@ -146,49 +154,82 @@
+
+/obj/machinery/atmospherics/miner/gas_portal/moles_to_spawn(var/turf/open/turf)
+	//Releases exactly enough gas to reach the target pressure.
+	var/datum/gas_mixture/G = turf.return_air()
```
name the var not `turf`

Cyberboss commented on code/modules/atmospherics/machinery/other/miner.dm:
```diff
@@ -146,49 +154,82 @@
+
+/obj/machinery/atmospherics/miner/gas_portal/repressurizer
+	name = "Bluespace atmosphere restorer"
+	desc = "A miniature bluespace portal used to siphon the atmosphere from a backwater planet so that you may refill a depresurized area on the station. This portal will dissipate one minute after creation."
```
remove `on the station`

Commit:
Small optimizations and grammar corrections.

Commit:
Fixes compilation breaking typo.

optimumtact commented:
@KorPhaeron I don't like this, I think atmos has enough tools to repressurize areas as it is and they don't need these.

lordpidey commented:
Why stale?  It's feature complete, and I have addressed all change requests.

optimumtact commented:
because no maintainer has touched for a while, which likely means they dont' want to merge it.

this way someone has to address it or it gets closed

KorPhaeron commented:
@optimumtact 

I'm not a very big atmos person, magic handwavey stuff like this or tiny fans doesn't bother me but I know it bothers people like @Cyberboss 

@Cyberboss @as334 do you have opinions on this?

Cyberboss commented on code/modules/atmospherics/machinery/other/miner.dm:
```diff
@@ -132,8 +138,10 @@
+	var/moles_spawned = moles_to_spawn(O)
+	for(var/gas in gas_mixture)
+		merger.assert_gas(gas)
+		merger.gases[gas][MOLES] = moles_spawned * gas_mixture[gas]
```
*\**?????

lordpidey replied:
Gas mixture is an associative list, that maps gas type to relative ratio,  for example, the airmix is `list("n2" = N2STANDARD, "o2" = O2STANDARD)`,

Thus If moles_to_spawn() is 5, it will add about 4 moles of N2, and 1 mole of O2.

I agree, the syntax is a bit confusing, especially since when you do an enhanced for loop over an associative list, it iterates over the keys rather than the values of the list, but that's a byond quirk.

I'll make a commit that clarifies this in a comment.

Commit:
Adds clarifying comment to confusing code section.

Commit:
Small optimization.

lordpidey commented:
Okay, I've added a comment to clarify how a loop works (Since it relies on unintuitive byond behavior).

I've also changed an assert_gas to an add_gas call, since we know that it's being called on an empty gas datum, for a small optimization.

Commit:
AI's no longer get erroneous messages if they click a portal in an overpressurized room.

optimumtact commented:
@Cyberboss merging this is on you

KorPhaeron commented:
@Cyberboss 

Reopen this if you plan to merge it, it's been two weeks

KorPhaeron commented:
Cyberboss wants this open again

optimumtact commented:
He gets 48 hours and then I'm turbo dunking this into space

lordpidey commented:
Pinging @Cyberboss

optimumtact commented:
INTO SPACE

lordpidey commented:
@Cyberboss YOU HAVE BETRAYED MEEEEEeee......
